ot/dr.py

```diff
@@ -30,16 +30,25 @@ def dist(x1, x2):
     return x1p2.reshape((-1, 1)) + x2p2.reshape((1, -1)) - 2 * np.dot(x1, x2.T)
 
 
+def logsumexp(M, axis):
+    r"""Log-sum-exp reduction compatible with autograd (no numpy implementation)
+    """
+    amax = np.amax(M, axis=axis, keepdims=True)
+    return np.log(np.sum(np.exp(M - amax), axis=axis)) + np.squeeze(amax, axis=axis)
```

Comment:
Why are you using `keepdims=True` and then `squeeze`? It should it work without both?

Reply:
`M - amax` (inside of `exp`) won't work without `keepdims=True` when reducing on `axis=1`. As an example:
```
>>> M = np.random.randn(10, 20)
>>> amax = np.amax(M, axis=1)
>>> M - amax
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: operands could not be broadcast together with shapes (10,20) (10,) 
```
while
```
>>> amax = np.amax(M, axis=1, keepdims=True)
>>> M - amax
```
works fine.

Reply:
ok you are right